ipam/allocator.go

```diff
@@ -159,389 +211,401 @@ const (
-	key := subnetKey{addrSpace, subnetInfo.Subnet.String(), ""}
-	a.Lock()
-	a.subnets[key] = subnetInfo
-	a.Unlock()
-	err = a.writeToStore()
+	insert, err := a.updatePoolDBOnAdd(*k, ipr, nw, sw)
 	if err != nil {
+		return "", nil, nil, err
+	}
+	if err := a.writeToStore(); err != nil {
 		if _, ok := err.(types.RetryError); !ok {
-			return types.InternalErrorf("subnet configuration failed because of %s", err.Error())
+			return "", nil, nil, types.InternalErrorf("pool configuration failed because of %s", err.Error())
 		}
-		// Update to latest
 		if erru := a.readFromStore(); erru != nil {
-			// Restore and bail out
-			a.Lock()
-			delete(a.addresses, key)
-			a.Unlock()
-			return fmt.Errorf("failed to get updated subnets config from datastore (%v) after (%v)", erru, err)
+			return "", nil, nil, fmt.Errorf("failed to get updated pool config from datastore (%v) after (%v)", erru, err)
 		}
 		goto retry
 	}
-
-	// Insert respective bitmasks for this subnet
-	a.insertAddressMasks(key, subnetList)
-
-	return nil
+	return k.String(), aw, nil, insert()
 }
 
-// Create and insert the internal subnet(s) addresses masks into the address database. Mask data may come from the bitseq datastore.
-func (a *Allocator) insertAddressMasks(parentKey subnetKey, internalSubnetList []*net.IPNet) error {
-	ipVer := getAddressVersion(internalSubnetList[0].IP)
-	num := len(internalSubnetList)
-	ones, bits := internalSubnetList[0].Mask.Size()
-	numAddresses := 1 << uint(bits-ones)
-
-	for i := 0; i < num; i++ {
-		smallKey := subnetKey{parentKey.addressSpace, parentKey.subnet, internalSubnetList[i].String()}
-		limit := uint32(numAddresses)
+// ReleasePool releases the address pool identified by the passed id
+func (a *Allocator) ReleasePool(poolID string) error {
+	k := SubnetKey{}
+	if err := k.FromString(poolID); err != nil {
+		return types.BadRequestErrorf("invalid pool id: %s", poolID)
+	}
 
-		if ipVer == v4 && i == num-1 {
-			// Do not let broadcast address be reserved
-			limit--
+retry:
+	remove, err := a.updatePoolDBOnRemoval(k)
+	if err != nil {
+		return err
+	}
+	if err = a.writeToStore(); err != nil {
+		if _, ok := err.(types.RetryError); !ok {
+			return types.InternalErrorf("pool (%s) removal failed because of %v", poolID, err)
 		}
-
-		// Generate the new address masks. AddressMask content may come from datastore
-		h, err := bitseq.NewHandle(dsDataKey, a.getStore(), smallKey.String(), limit)
-		if err != nil {
-			return err
+		if erru := a.readFromStore(); erru != nil {
+			return fmt.Errorf("failed to get updated pool config from datastore (%v) after (%v)", erru, err)
 		}
+		goto retry
+	}
 
-		if ipVer == v4 && i == 0 {
-			// Do not let network identifier address be reserved
-			h.Set(0)
-		}
+	return remove()
+}
 
-		a.Lock()
-		a.addresses[smallKey] = h
-		a.Unlock()
+func (a *Allocator) parsePoolRequest(addressSpace, pool, subPool string, v6 bool) (*SubnetKey, *net.IPNet, *net.IPNet, *net.IPNet, *AddressRange, error) {
```

Comment:
Now I really have to hate golang for allowing us to return 6 values :)
I understand that this function is complex. But the signature need not be.
Can you please wrap these return values into a structure and return a pointer ?
That will make the code much more readable compared to `nw, aw, sw *net.IPNet` ;)